src/miner.cpp

```diff
@@ -832,93 +838,100 @@ namespace pos {
-            }
-            if (status == Staker::Status::initWaiting) {
-                LogPrintf("ThreadStaker (%s): waiting init...\n", operatorName);
-            }
-            if (status == Staker::Status::stakeWaiting) {
-                LogPrint(BCLog::STAKING, "ThreadStaker (%s): Staked, but no kernel found yet.\n", operatorName);
+        for (auto it = args.begin(); it != args.end(); ) {
+            const auto& arg = *it;
+            const auto operatorName = arg.operatorID.GetHex();
+
+            pos::Staker staker;
+
+            try {
+                auto status = staker.init(chainparams);
+                if (status == Staker::Status::stakeReady) {
+                    status = staker.stake(chainparams, arg);
+                }
+                if (status == Staker::Status::error) {
+                    LogPrintf("ThreadStaker: (%s) terminated due to a staking error!\n", operatorName);
+                    it = args.erase(it);
+                    continue;
+                }
+                if (status == Staker::Status::minted) {
+                    LogPrintf("ThreadStaker: (%s) minted a block!\n", operatorName);
+                    nMinted[arg.operatorID]++;
+                }
+                if (status == Staker::Status::initWaiting) {
+                    LogPrintf("ThreadStaker: (%s) waiting init...\n", operatorName);
+                }
+                if (status == Staker::Status::stakeWaiting) {
+                    LogPrint(BCLog::STAKING, "ThreadStaker: (%s) Staked, but no kernel found yet.\n", operatorName);
+                }
+                if (status == Staker::Status::criminalWaiting) {
+                    LogPrint(BCLog::STAKING, "ThreadStaker: (%s) Potential criminal block tried to create.\n", operatorName);
+                }
             }
-            if (status == Staker::Status::criminalWaiting) {
-                LogPrint(BCLog::STAKING, "ThreadStaker (%s): Potential criminal block tried to create.\n", operatorName);
+            catch (const std::runtime_error &e) {
+                LogPrintf("ThreadStaker: (%s) runtime error: %s\n", e.what(), operatorName);
+
+                // Could be failed TX in mempool, wipe mempool and allow loop to continue.
+                mempool.clear();
             }
-        }
-        catch (const std::runtime_error &e) {
-            LogPrintf("ThreadStaker (%s): runtime error: %s\n", e.what(), operatorName);
 
-            // Could be failed TX in mempool, wipe mempool and allow loop to continue.
-            mempool.clear();
-        }
+            auto& tried = nTried[arg.operatorID];
+            tried++;
+
+            if ((arg.nMaxTries != -1 && tried >= arg.nMaxTries)
+            || (arg.nMint != -1 && nMinted[arg.operatorID] >= arg.nMint)) {
+                it = args.erase(it);
+                continue;
+            }
 
-        nTried++;
-        if (trying() && notDone()) {
             std::this_thread::sleep_for(std::chrono::milliseconds(900));
```

Comment:
@bvbfan I think sleep here is sleeping between every MN, we do not want to sleep here every 900 ms! This needs to be moved out of MN loop